staging/src/k8s.io/apiextensions-apiserver/test/integration/testserver/resources.go

```diff
@@ -309,24 +315,28 @@ func isWatchCachePrimed(crd *apiextensionsv1beta1.CustomResourceDefinition, dyna
 		return false, err
 	}
 
-	noxuWatch, err := resourceClient.Watch(metav1.ListOptions{ResourceVersion: createdInstance.GetResourceVersion()})
-	if err != nil {
-		return false, err
-	}
-	defer noxuWatch.Stop()
-
-	select {
-	case watchEvent := <-noxuWatch.ResultChan():
-		if watch.Error == watchEvent.Type {
-			return false, nil
+	for _, v := range versions {
+		noxuWatch, err := resourceClientForVersion(crd, dynamicClientSet, ns, v).Watch(
+			metav1.ListOptions{ResourceVersion: createdInstance.GetResourceVersion()})
+		if err != nil {
+			return false, err
 		}
-		if watch.Deleted != watchEvent.Type {
-			return false, fmt.Errorf("expected DELETE, but got %#v", watchEvent)
+		defer noxuWatch.Stop()
+
+		select {
+		case watchEvent := <-noxuWatch.ResultChan():
```

Comment:
what I still miss is an explaination (comment? reference to axiom) that a `rv=createdInstance.GetResourceVersion()` will drain the watch cache for a following `rv=0` watch. All this here is based on that behaviour. And it is far from obvious.

Reply:
Will add a comment. I couldn't find an official document describing the behaviour. I looked at the code and tested it.

Reply:
Also getResourceVersion does not drain anything. Abwatch cache has a items that has not been sent to any watch. When a watch is created with rc=0, it will get all of those. If those events drained by being sent to another watch (or skipped because the watch is requesting a more recent rc and actually getting something), they will not be sent to watch with rc=0 in future.

Reply:
Basically when you get an event from a watch channel by any client at t0, and another client creates a watch at t1 with rc=0, it may only get events after the event sent at t0. I said may because it is a circular buffer and there is no guarantee that all events after t0 will be delivered. Watch uses one cache per resource/kind/version so I guess it was tricky not to do that (some other concurrent watches may miss events if we drain cache on new watches with rc=0. Not impossible to fix but maybe tricky). Fixing that behaviour is not in the scope of this PR and I'm not sure it ever get fixed. A watch with rc not zero is predictable and recommend if we want predictibility.